libcentrifugo/engineredis.go

```diff
@@ -427,11 +446,65 @@ func (e *RedisEngine) runPubSub() {
 	}
 }
 
-func (e *RedisEngine) publish(chID ChannelID, message []byte) (bool, error) {
+func (e *RedisEngine) publish(chID ChannelID, message []byte, opts *publishOpts) error {
 	conn := e.pool.Get()
 	defer conn.Close()
-	numSubscribers, err := redis.Int(conn.Do("PUBLISH", chID, message))
-	return numSubscribers > 0, err
+
+	var err error
+	if opts == nil {
+		// just publish message into channel.
+		_, err = e.pubScript.Do(conn, chID, message, "0")
```

Comment:
Hmm `NewScript` call above specified 3 key arguments but we only pass 1 here.

I think you are mixing and matching "keys" and "args" to the script which is not ideal. Redis docs say the "KEYS" arguments should "represent Redis key names."

We should be stricter, I suggest:
 - KEYS[1] is the channel's history list key 
 - ARGS[1] is the channel id for `PUBLISH`
 - ARGS[2] is the message payload
 - ARGS[3] is the `HistorySize` (0 means don't save history)
 - ARGS[4] is the `HistoryLifetime`
 - ARGS[5] is the `HistoryDropInactive` (0 or 1)

Then we can call it like this:

```go
// Initialize script with NewScript(1)

if  opts == nil {
    e.pubScript.Do(conn, e.getHistoryKey(chID), chID, message, 0, 0, 0);
} else {
    e.pubScript.Do(conn, e.getHistoryKey(chID), chID, message, opts.HistorySize, opts.HistoryLifetime, opts.HistoryDropInactive);
}
```

Then the script itself becomes:

```lua
local histSize = tonumber(ARGV[3])
local histLifetime = tonumber(ARGV[4])
local histDropInactive = ARGV[5]
local n = redis.call("publish", ARGV[1], ARGV[2])
if histSize > 0 then
  local m = 0
  if histDropInactive == "1" and n == 0 then
    m = redis.call("lpushx", KEYS[1], ARGV[2])
  else
    m = redis.call("lpush", KEYS[1], ARGV[2])
  end
  if m > 0 then
    redis.call("ltrim", KEYS[1], 0, histSize)
    redis.call("expire", KEYS[1], histLifetime)
  end
end
return n
```